### PR TITLE
Add autocomplete for bash and zsh in lokoctl

### LINKED ISSUE #18: Add Bash/zsh autocompletion
Cobra [supports](https://github.com/spf13/cobra#generating-shell-completions) generating autocomplete files for Bash and zsh. We could leverage this to make working with lokoctl nicer.

### PR DESCRIPTION
**Pull request by knrt10**
Fixes #18

Signed-off-by: knrt10 <kautilya@kinvolk.io>

### CODE REVIEW

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:88]: Why not use `GenZshCompletion()` completion for zsh?

Comment (knrt10) [reply cli/cmd/completion.go:88]: I was using that before, but it flags were preferred over nouns. https://github.com/spf13/cobra/issues/1189

Comment (marckhouzam) [reply cli/cmd/completion.go:88]: Cobra 1.0.0 does not support `ValidArgsFunction` for `zsh`.  You will need to wait for the next release, or you can point to the master branch which has the full support required.

Comment (invidian) [reply cli/cmd/completion.go:88]: @knrt10 what's the diff between the version generated by cobra and this one? Can we maybe patch it dynamically to avoid hardcoding it?

File: docs/cli/lokoctl_completion_zsh.md
Comment (BrainBlasted) [inline docs/cli/lokoctl_completion_zsh.md:3]: Here it's autocompletions, and below it's autocompletion. In keeping them consistent, I think it should be autocompletion here?

File: cli/cmd/completion.go
Comment (BrainBlasted) [inline cli/cmd/completion.go:48]: Here as well, in addition to other places in the documentation.

File: cli/cmd/completion.go
Comment (marckhouzam) [reply cli/cmd/completion.go:88]: > what's the diff between the version generated by cobra and this one? Can we maybe patch it dynamically to avoid hardcoding it?

If you allow me to answer this one.
On cobra's master branch, there was a complete redesign of the zsh completion solution.  
It was a major change as you can see from:
https://github.com/spf13/cobra/pull/1070

Until it is released, the zsh completion (of cobra 1.0) is incomplete.

The alternative used in this pr is what `helm` and `kubectl` currently use for zsh, until the next cobra release.  This temporary solution is based on bash completion which is fully supported in cobra 1.0.

Comment (invidian) [reply cli/cmd/completion.go:88]: @marckhouzam thanks for the reply! And when do you plan 1.0 release roughly?

**Comment by knrt10** [comment]
@invidian, so I think, currently go for this untill cobra releases it's new updates. Do you agree?

**Comment by knrt10** [comment]
For 0.4 we can release it with this, if by 0.5 cobra releases their new updates, we can change it to `GenZshCompletion`

File: cli/cmd/completion.go
Comment (marckhouzam) [reply cli/cmd/completion.go:88]: @invidian I'm not a maintainer of the Cobra project but the suggestion of a new release has been mentioned. I'm hopeful it will happen in the next few weeks, but no guarantee.

**Comment by invidian** [comment]
>For 0.4 we can release it with this, if by 0.5 cobra releases their new updates, we can change it to GenZshCompletion

Or we could do bash completion for 0.4 and zsh completion for the next release? What do you think @johananl? Or we could postpone this completely.

**Comment by knrt10** [comment]
As mentioned in https://github.com/kinvolk/lokomotive/pull/880#discussion_r479638982 we don't know when the next cobra release will be, but if we do want to use `GenZshCompletion` one alternative is to use, `ValidArgs` instead of `ValidArgsFunction` and provide arguments explicitly for both cluster apply and delete.

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:29]: Maybe we can use string literal here to make the formatting simpler? https://golang.org/ref/spec#String_literals

Also, it seems we could make that a function, to de-duplicate parts of the text.

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:52]:
```suggestion
		ValidArgsFunction: noCompletions, // Disable file completion.
```

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:46]: Is this possible to mark parameter as not used?
```suggestion
func newCompletionCmd(_ io.Writer) *cobra.Command {
```

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:89]: I think we could move those fragments to constants.

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:223]:
```suggestion
	if _, err := out.Write([]byte(zshInitialization)); err != nil {
	  fmt.Errorf("writing zsh initialization: %w", err)
	}
```

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:228]:
```suggestion
	if err := runCompletionBash(out, cmd); err != nil {
		return fmt.Errorf("running bash completion: %w, err)
	}
```

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:239]:
```suggestion
	if _, err = out.Write([]byte(zshTail)); err != nil {
		return fmt.Errorf("writing zsh tail: %w", err)
	}
```

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:51]: Do we really need to depend on Helm code for that?

**Review by invidian — CHANGES_REQUESTED** [review]
Some nits, otherwise LGTM :+1:

File: cli/cmd/completion.go
Comment (knrt10) [reply cli/cmd/completion.go:46]: But we need this writer. It is passed onto `runCompletionZsh` and `runCompletionBash`

Comment (invidian) [reply cli/cmd/completion.go:46]: Ah, sorry, I missed that :+1:

File: docs/cli/lokoctl.md
Comment (invidian) [inline docs/cli/lokoctl.md:24]: I'd actually put this all in a single commit, so if one runs the build per commit, it stays atomic.

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:30]: Why do we need indentation here and extra newline at the beginning?

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:257]: Please consolidate lines which only assigns `err` with `if` statement lines.
```suggestion
	if _, err := out.Write([]byte(zshInitialization)); err != nil {
```

**Review by invidian — CHANGES_REQUESTED** [review]

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:29]: Looking at current help message format:
```
Usage:
  lokoctl completion [command]

Examples:
# Installing bash completion on macOS using homebrew
## If running Bash 3.2 included with macOS
  brew install bash-completion
## or, if running Bash 4.1+
  brew install bash-completion@2
## You need to add the completion to your completion directory
  lokoctl completion bash > /usr/local/etc/bash_completion.d/lokoctl && exec $SHELL

# Installing bash completion on Linux
## If bash-completion is not installed on Linux, please install the 'bash-completion' package
## via your distribution's package manager.
## You need to add the completion to your completion directory
  lokoctl completion bash > /etc/bash_completion.d/lokoctl && exec $SHELL

# Load the lokoctl completion code for bash into the current shell
  source <(lokoctl completion bash)
# Load the lokoctl completion code for zsh into the current shell
  source <(lokoctl completion zsh)

# Set the lokoctl completion code for zsh to autoload on startup
  lokoctl completion zsh > "${fpath[1]}/_lokoctl" && exec $SHELL

Available Commands:
  bash        Generate autocompletion script for bash
  zsh         Generate autocompletion script for zsh

Flags:
  -h, --help   help for completion
```

It seems example text could be indented with 2 spaces and should not use `#` characters. Can you re-format it @knrt10 to some better format?

**Review by invidian — CHANGES_REQUESTED** [review]

File: cli/cmd/completion.go
Comment (johananl) [reply cli/cmd/completion.go:29]: +1. Also, we should pay more attention to the following in our codebase in general and especially in text the users see:

- Punctuation.
  - Some lines in the examples text end in periods whereas others don't.
- Capitalization.
  - We use both `Bash` and `bash`. We should use `Bash`, and anyway we should be consistent.
  - Some lines begin with a capital letter whereas others don't.
- Consistent tone and style.
  - We should be consistent about using the imperative tone and avoiding "please" as in "please do this and that".
  - In one instance we use `To load completions in your current shell session:` and then immediately following we use `To load completions for every new session, execute once:`.
- Correct English.
  - I'm fairly certain the following sentence has some words missing in it (probably `that's` before `included`): `If running Bash 3.2 included with macOS`.

**Comment by knrt10** [comment]
@johananl can you please review again

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:26]:
```suggestion
Generate autocompletion code for lokoctl for the specified shell (bash or zsh).
```

I find "code" better than "script" in this context. In any case, if we use "script" then it should be "Generate **an** autocompletion script..." since "script" is countable (unlike "code" in this context).

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:35]:
```suggestion
  # Generate a Bash completion file and load it for every shell.
```

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:42]: Why use `###`? Is this some sort of convention?

Comment (johananl) [inline cli/cmd/completion.go:42]:
```suggestion
const bashCompDesc = `  ### If running Bash 3.2 that is included with macOS, install bash completion using Homebrew.
```

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:26]: We should consistently capitalize "Bash" as `Bash`.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:56]:
```suggestion
const zshCompDesc = `  Generate an autocompletion script for lokoctl for the zsh shell.
```

Comment (johananl) [inline cli/cmd/completion.go:56]: We don't have an equivalent line in the long description for `lokoctl completion bash`.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:58]: We seem to be using "autocompletion" and "completion" randomly. We should settle on one and use it consistently. Same for "code" vs "script" (I prefer "code").

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:78]: Does this text come from upstream? If so, please ignore this comment as well as similar comments in this file.

- The comment should probably begin with an upper case letter.
- The last sentence doesn't have a period.
- I find the phrasing in this comment confusing.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:104]: Missing period.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:110]: - Caps.
- Period.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:115]: Period.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:133]: - Caps.
- Period.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:170]: - Caps.
- Period.

File: cli/cmd/completion.go
Comment (johananl) [inline cli/cmd/completion.go:163]: Period.

**Review by johananl — COMMENTED** [review]
Thanks @knrt10.

I'm still seeing small inconsistencies. I've highlighted some of them inline. We also seem to be using multiple variations of "completion/autocompletion" and "script/code". Let's settle on one phrasing :wink:

File: cli/cmd/completion.go
Comment (knrt10) [reply cli/cmd/completion.go:42]: For docs, otherwise, they look wierd

File: cli/cmd/completion.go
Comment (knrt10) [reply cli/cmd/completion.go:78]: Yes, it's from upstream.

**Comment by knrt10** [comment]
Updated

File: cli/cmd/completion.go
Comment (invidian) [inline cli/cmd/completion.go:44]: I'd be consistent with those comments.
```suggestion
  ### If running Bash 3.2 that is included with macOS, install Bash completion using Homebrew.
```

**Review by invidian — CHANGES_REQUESTED** [review]
Just one nit, otherwise looks fine to me.

File: cli/cmd/completion.go
Comment (iaguis) [inline cli/cmd/completion.go:42]: This looks pretty ugly in the documentation.

Doesn't it make more sense to have this as `Example` and add a small `Description` like we do for the `completion` command? The instructions below feel more like examples and it would make it more consistent too. We can also remove the triple `#`.

File: cli/cmd/completion.go
Comment (iaguis) [inline cli/cmd/completion.go:58]: See my previous comment.

File: cli/cmd/completion.go
Comment (iaguis) [inline cli/cmd/completion.go:179]: Can we add a todo somewhere to get rid of this when Cobra is released?

**Review by iaguis — CHANGES_REQUESTED** [review]
Some small comments.

**Review by iaguis — APPROVED** [review]
lgtm

**Review by invidian — APPROVED** [review]
LGTM

**Comment by knrt10** [comment]
tested for both zsh and bash. Merging now

**Pull request merged**
